### PR TITLE
An example showing how to do clustering on the RandomTreesEmbedding outputs with modified distance metrics to solve Issue #3

### PR DESCRIPTION
**Pull request by ShumingXu**
#### Reference Issues/PRs

This pull request tries to settle the problem raised by Issue #3 

#### What does this implement/fix? Explain your changes.
An example is added to showcase how different distance metrics related to tree structure can be generated from the outputs of Random Trees Embedding and be used on clustering. 
The distance metrics are: 1) depth of nearest common ancestor; 2) length of shortest path; 3) proximity matrix. The code is implemented on Iris dataset as a simple example. 

#### Any other comments?
I have a tutorial jupyter notebook written but I don't know where exactly I should put it. Further feedback is welcome.

### CODE REVIEW

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+clf = clf.fit(samples)
+
+# compute the first distance metrics, which is 
+# one over the depth of the nearest common ancestor
+# averaged over all estimators
+
+forest = clf.estimators_
+d = np.zeros([n_sample, n_sample])
+
+for k in range(n_estimator):
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: You should write functions for each of the three methods you are comparing, probably makes sense for them to take in an estimator and output a distance matrix or something like that

Comment (bdpedigo) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: This will just be more clear to the reader

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+print("predict\n", predict_labels)
+print("Adjusted Rand Score:", score)
+
+# compute the second distance metrics, which is 
+# length of shortest path between leaf nodes
+# averaged over all estimators
+
+d = np.zeros([n_sample, n_sample])
+
+for k in range(n_estimator):
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: same comment about writing a function instead

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+print("labels\n", labels)
+print("predict\n", predict_labels)
+print("Adjusted Rand Score:", score)
+
+# compute the third distance metrics, which is 
+# the probability of two samples landing into
+# the same leaf node in an estimator
+
+Y = clf.transform(samples)
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: same comment about writing a function instead

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+prob = 1 - prob / Y.shape[1]
+        
+# plot the heatmap
+
+plt.subplot(1, 3, 3)
+plt.imshow(prob)
+
+# use agglomerative clustering to classify the output
+
+cluster = AgglomerativeClustering(n_clusters=3, affinity="precomputed", linkage="average")
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:181]: why does this one use a different linkage than the others?

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+from sklearn import tree
+from sklearn import datasets
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn.ensemble import RandomTreesEmbedding
+
+# use the Iris dataset as a simple example
+
+iris = datasets.load_iris()
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:122]: I would like to see this on more than just iris, lets do some simple simulation examples? How about the one from here for a start? https://scikit-learn.org/stable/auto_examples/ensemble/plot_random_forest_embedding.html#sphx-glr-auto-examples-ensemble-plot-random-forest-embedding-py

Or other examples from sklearn or jovo's papers would be a good place to start. If you have a reason to do a simulation other than those that you think would be more informative, let us know and we can discuss

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+plt.subplot(1, 3, 1)
+plt.imshow(d_Nearest_Common_Ancestor)
+
+# use agglomerative clustering to classify the output
+
+from sklearn.cluster import AgglomerativeClustering
+from sklearn.metrics import adjusted_rand_score
+
+cluster = AgglomerativeClustering(n_clusters=3, affinity="precomputed", linkage="complete")
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:161]: I'm curious, how did you choose `linkage='complete'`

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+d_Nearest_Common_Ancestor = [1/x for x in d]
+
+# plot the heatmap 
+
+plt.subplot(1, 3, 1)
+plt.imshow(d_Nearest_Common_Ancestor)
+
+# use agglomerative clustering to classify the output
+
+from sklearn.cluster import AgglomerativeClustering
```
Comment (bdpedigo) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: imports at top

**Review by bdpedigo — CHANGES_REQUESTED** [review]

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 5 above)
Comment (ShumingXu) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:122]: I'm thinking that this will be only an example of how to implement the code, and write a saparate  jupyter notebook that do the demonstration. I have one here but I don't know where I should put it and the notebook needs some change. Do you have any idea?

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 6 above)
Comment (ShumingXu) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:161]: It's through trial. Turns out using the maximum of the distance gives better results.

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 1 above)
Comment (ShumingXu) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: Makes sense. Changed accordingly.

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 4 above)
Comment (ShumingXu) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:181]: Turns out it gives better result.

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 7 above)
Comment (ShumingXu) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py]: Changed.

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 4 above)
Comment (bdpedigo) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:181]: don't tell me, tell the next person who will read the notebook by updating the description/code accordingly. 

My point is mainly that since you haven't done any systematic parameter tuning (at least in the notebook that you are showing) then it could look like you are only picking certain results that tell the story you want to tell

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 5 above)
Comment (bdpedigo) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:122]: > I'm thinking that this will be only an example of how to implement the code, and write a saparate jupyter notebook that do the demonstration. 

That's fine

> I have one here but I don't know where I should put it and the notebook needs some change. Do you have any idea?

You mean you don't know where to put it in sklearn? I wouldn't worry about that much for now, just put it wherever you put this file and we can figure that out later. Was that your question?

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 6 above)
Comment (bdpedigo) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:161]: see my comment above about parameter tuning - unclear what the default should be here

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py (hunk 5 above)
Comment (ShumingXu) [reply examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:122]: I changed the notebook according to the feedback today and now it's uploaded.

**Comment by bdpedigo** [comment]
- the colors for plotting the data are a bit harsh for me, can you consider using one of the default color maps like `deep` in matplotlib?
- can remove x and y ticks and tick labels when you are plotting the data, I think. They don't add anything to the story
- add a `plt.tight_layout()` at the very end of your plotting, see if that uses space better
- remove empty cell at end 
- increase size of titles in last plot

Notebook is a lot clearer I think, nice work!

**Comment by bdpedigo** [comment]
maybe the first plot of the data is unnecessary, what do you think?

**Comment by ShumingXu** [comment]
> * the colors for plotting the data are a bit harsh for me, can you consider using one of the default color maps like `deep` in matplotlib?
> * can remove x and y ticks and tick labels when you are plotting the data, I think. They don't add anything to the story
> * add a `plt.tight_layout()` at the very end of your plotting, see if that uses space better
> * remove empty cell at end
> * increase size of titles in last plot
> 
> Notebook is a lot clearer I think, nice work!

I changed the notebook according to the advice. But I didn't find a colormap called 'deep' so I used 'viridis' instead.

**Review by bdpedigo — APPROVED** [review]

**Comment by bdpedigo** [comment]
well done @ShumingXu !

File: examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        for i in range(n_sample):
+            for j in range(n_sample):
+                sample_ids = [i, j]
+                d[i, j] = d[i, j]+1/(path[sample_ids].sum(axis=0) == len(sample_ids)).sum()
+            
+
+    d = d / n_estimator
+    d_Nearest_Common_Ancestor = [1/x for x in d]
+    
+    return d_Nearest_Common_Ancestor
```
Comment (MaiRajborirug) [inline examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py:65]: Should add `d_Nearest_Common_Ancestor = np.asarray(d_Nearest_Common_Ancestor)` before `return`

**Review by MaiRajborirug — COMMENTED** [review]
I suggest the output of distance matrix function `NCA(forest, samples)` to be <class 'numpy.ndarray'> instead of <class 'list'>. It will be easier to compute outlier prediction on the distance matrices
